noir-projects/aztec-nr/authwit/src/auth.nr

```diff
@@ -311,8 +311,17 @@ pub fn compute_authwit_message_hash_from_call<let N: u32>(
  *
  * @param args The arguments to hash
  */
-pub fn compute_inner_authwit_hash<let N: u32>(args: [Field; N]) -> Field {
-    poseidon2_hash_with_separator(args, GENERATOR_INDEX__AUTHWIT_INNER)
+pub fn compute_inner_authwit_hash(args: [Field; 3]) -> Field {
+    poseidon2_hash_3_with_separator(args, GENERATOR_INDEX__AUTHWIT_INNER as Field)
+}
+
+fn poseidon2_hash_3_with_separator<T>(args: [Field; 3], separator: T) -> Field where T: ToField {
+    let two_pow_64 = 18446744073709551616;
+    let iv : Field = 4 * two_pow_64;
+    let mut state = [separator.to_field(), args[0], args[1], iv];
+    state = std::hash::poseidon2_permutation(state, 4);
+    state[0] = state[0] + args[2];
+    std::hash::poseidon2_permutation(state, 4)[0] as Field
 }
```

Comment:
I don't know how poseidon works so this is just arcane magic to me. Could you add a small explanation. Even a brief comment explaning that this is a specialized version for 3 fields which results in ??? (fewer constraints?) would be nice.

We should also have some tests that verify that this implementation is equivalent to `poseidon2_hash_with_separator`.